File: STRM/src/main/java/hcs/offloading/strm/RoIExtractor.java
package hcs.offloading.strm;

import android.graphics.Bitmap;
import android.graphics.Rect;
import android.util.Log;
import android.util.Pair;

import org.opencv.android.OpenCVLoader;
import org.opencv.android.Utils;
import org.opencv.core.Core;
import org.opencv.core.Mat;
import org.opencv.core.MatOfByte;
import org.opencv.core.MatOfFloat;
import org.opencv.core.MatOfPoint;
import org.opencv.core.MatOfPoint2f;
import org.opencv.core.Point;
import org.opencv.core.Size;
import org.opencv.core.TermCriteria;
import org.opencv.imgproc.Imgproc;
import org.opencv.video.Video;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import hcs.offloading.strm.config.RoIExtractorConfig;
import hcs.offloading.strm.datatypes.BoundingBox;
import hcs.offloading.strm.datatypes.Frame;
import hcs.offloading.strm.datatypes.RoI;

public class RoIExtractor {
    private static final String TAG = RoIExtractor.class.getName();

    static {
        if (!OpenCVLoader.initDebug()) Log.e("OpenCV", "Unable to load OpenCV!");
        else Log.d("OpenCV", "OpenCV loaded Successfully");
    }

    private final RoIExtractorConfig mConfig;
    private final Size mTargetSize;

    public RoIExtractor(RoIExtractorConfig config) {
        mConfig = config;
        mTargetSize = new Size(mConfig.EXTRACTION_RESIZE_WIDTH, mConfig.EXTRACTION_RESIZE_HEIGHT);
    }

    public boolean useOpticalFlowRoIs() {
        return mConfig.OF_ROI;
    }

    public void process(Pair<Pair<Frame, Frame>, List<BoundingBox>> item) {
        Frame prevFrame = item.first.first;
        Frame currFrame = item.first.second;

        List<RoI> rois = new ArrayList<>();
        if (mConfig.OF_ROI) {
            List<BoundingBox> prevResults = item.second.stream()
                    .filter(box -> box.confidence > mConfig.OPTICAL_FLOW_ROI_CONFIDENCE_THRESHOLD)
                    .collect(Collectors.toList());
            List<RoI> opticalFlowRoIs = getOpticalFlowRoIs(prevFrame, currFrame, prevResults, mTargetSize);
            currFrame.setOpticalFlowRoIs(opticalFlowRoIs);
            rois.addAll(opticalFlowRoIs);
        }
        if (mConfig.PD_ROI) {
            rois.addAll(getPixelDiffRoIs(prevFrame, currFrame, mTargetSize));
        }
        if (mConfig.MERGE_ROI) {
            rois = mergeSingleFrameRoIs(currFrame, rois, mConfig.MERGE_THRESHOLD);
        }
        currFrame.setRoIs(rois);
    }

    private static List<RoI> mergeSingleFrameRoIs(Frame frame, List<RoI> rois, float mergeThreshold) {
        List<RoI> mergedRoIs = new ArrayList<>(rois);
        while (true) {
            RoI originalRoI0 = null;
            RoI originalRoI1 = null;
            RoI mergedRoI = null;
            for (int i = 0; i < mergedRoIs.size(); i++) {
                for (int j = i + 1; j < mergedRoIs.size(); j++) {
                    RoI roi0 = mergedRoIs.get(i);
                    RoI roi1 = mergedRoIs.get(j);
                    float intersection = STRMUtils.box_intersection(roi0.location, roi1.location);
                    if (intersection / roi0.getArea() < mergeThreshold
                            && intersection / roi1.getArea() < mergeThreshold) {
                        continue;
                    }
                    int newLeft = Math.min(roi0.location.left, roi1.location.left);
                    int newTop = Math.min(roi0.location.top, roi1.location.top);
                    int newRight = Math.max(roi0.location.right, roi1.location.right);
                    int newBottom = Math.max(roi0.location.bottom, roi1.location.bottom);
                    if (newLeft >= newRight || newTop >= newBottom) {
                        continue;
                    }
                    originalRoI0 = roi0;
                    originalRoI1 = roi1;
                    Rect newLocation = new Rect(newLeft, newTop, newRight, newBottom);
                    RoI.Type roiType = roi0.type.equals(RoI.Type.OF) || roi1.type.equals(RoI.Type.OF)
                            ? RoI.Type.OF
                            : RoI.Type.PD;
                    String roiLabel = roi0.labelName == null || roi1.labelName == null
                            || !roi0.labelName.equals(roi1.labelName)
                            ? null : roi0.labelName;
                    mergedRoI = new RoI(frame, newLocation, roiType, roiLabel,
                            Math.min(Math.max(roi0.location.width(), roi0.location.height()),
                                    Math.max(roi1.location.width(), roi1.location.height())));
                    break;
                }
                if (mergedRoI != null) {
                    break;
                }
            }
            if (mergedRoI == null) {
                break;
            }
            mergedRoIs.remove(originalRoI0);
            mergedRoIs.remove(originalRoI1);
            mergedRoIs.add(mergedRoI);
        }
        return mergedRoIs;
    }

    private static List<RoI> getOpticalFlowRoIs(
            Frame prevFrame, Frame currFrame, List<BoundingBox> boundingBoxes, Size targetSize) {
        final int width = currFrame.bitmap.getWidth();
        final int height = currFrame.bitmap.getHeight();

        List<Rect> boundingRects = boundingBoxes.stream()
                .map(bbx -> bbx.location)
                .collect(Collectors.toList());

        List<RoI> opticalFlowRoIs = new ArrayList<>();
        if (!boundingBoxes.isEmpty()) {
            int[][] shifts = getBoundingBoxShifts(
                    prevFrame.bitmap, currFrame.bitmap, boundingRects, targetSize);
            for (int boxIndex = 0; boxIndex < boundingBoxes.size(); boxIndex++) {
                int[] shift = shifts[boxIndex];
                BoundingBox box = boundingBoxes.get(boxIndex);
                Rect loc = box.location;
                int newLeft = Math.max(0, loc.left + shift[0]);
                int newTop = Math.max(0, loc.top + shift[1]);
                int newRight = Math.min(width, loc.right + shift[0]);
                int newBottom = Math.min(height, loc.bottom + shift[1]);
                if (newLeft < newRight && newTop < newBottom) {
                    opticalFlowRoIs.add(new RoI(
                            currFrame, new Rect(newLeft, newTop, newRight, newBottom),
                            RoI.Type.OF, box.labelName));
                }
            }
        }
        return opticalFlowRoIs;
    }

    private static int[][] getBoundingBoxShifts(
            Bitmap prevImage, Bitmap currImage, List<Rect> boundingBoxes, Size targetSize) {
        Mat prevMat = new Mat();
        Mat currMat = new Mat();
        MatOfPoint2f p0 = new MatOfPoint2f();
        MatOfPoint2f p1 = new MatOfPoint2f();
        MatOfByte status = new MatOfByte();
        MatOfFloat err = new MatOfFloat();

        Utils.bitmapToMat(prevImage, prevMat);
        Utils.bitmapToMat(currImage, currMat);

        Imgproc.cvtColor(prevMat, prevMat, Imgproc.COLOR_BGR2GRAY);
        Imgproc.cvtColor(currMat, currMat, Imgproc.COLOR_BGR2GRAY);

        Imgproc.resize(prevMat, prevMat, targetSize);
        Imgproc.resize(currMat, currMat, targetSize);

        List<Point> centroids = boundingBoxes.stream()
                .map(bbx -> new Point(
                        (float) bbx.centerX() * targetSize.width / currImage.getWidth(),
                        (float) bbx.centerY() * targetSize.height / currImage.getHeight()))
                .collect(Collectors.toList());
        p0.fromList(centroids);
        TermCriteria criteria = new TermCriteria(TermCriteria.COUNT + TermCriteria.EPS, 10, 0.03);
        Video.calcOpticalFlowPyrLK(prevMat, currMat, p0, p1, status, err, new Size(15, 15), 2, criteria);

        byte[] StatusArr = status.toArray();
        Point[] p1Arr = p1.toArray();

        int[][] shifts = new int[centroids.size()][2];
        for (int pointIdx = 0; pointIdx < centroids.size(); pointIdx++) {
            if (StatusArr[pointIdx] == 1) {
                shifts[pointIdx][0] = (int) ((p1Arr[pointIdx].x - centroids.get(pointIdx).x)
                        * currImage.getWidth() / targetSize.width);
                shifts[pointIdx][1] = (int) ((p1Arr[pointIdx].y - centroids.get(pointIdx).y)
                        * currImage.getHeight() / targetSize.height);
            } else {
                shifts[pointIdx][0] = 0;
                shifts[pointIdx][1] = 0;
            }
        }

        prevMat.release();
        currMat.release();
        p0.release();
        p1.release();
        status.release();
        err.release();
        return shifts;
    }

    private static List<RoI> getPixelDiffRoIs(Frame prevFrame, Frame currFrame, Size targetSize) {
        Mat prevMat = new Mat();
        Mat currMat = new Mat();

        Utils.bitmapToMat(prevFrame.bitmap, prevMat);
        Utils.bitmapToMat(currFrame.bitmap, currMat);

        Imgproc.cvtColor(prevMat, prevMat, Imgproc.COLOR_BGR2GRAY);
        Imgproc.cvtColor(currMat, currMat, Imgproc.COLOR_BGR2GRAY);

        Imgproc.resize(prevMat, prevMat, targetSize);
        Imgproc.resize(currMat, currMat, targetSize);

        Mat mat = new Mat();
        calculateDiffAndThreshold(prevMat, currMat, mat);
        cannyEdgeDetection(mat);

        ArrayList<MatOfPoint> contours = new ArrayList<>();
        Mat hierarchy = new Mat();
        Imgproc.findContours(mat, contours, hierarchy, Imgproc.RETR_EXTERNAL, Imgproc.CHAIN_APPROX_SIMPLE);
        List<Rect> locations = getBoxesFromContours(contours,
                currFrame.bitmap.getWidth(), currFrame.bitmap.getHeight(),
                (int) targetSize.width, (int) targetSize.height);
        List<RoI> rois = locations.stream()
                .map(location -> new RoI(currFrame, location, RoI.Type.PD, null))
                .collect(Collectors.toList());

        prevMat.release();
        currMat.release();
        hierarchy.release();
        mat.release();
        for (MatOfPoint contour : contours) {
            contour.release();
        }
        return rois;
    }

    private static void calculateDiffAndThreshold(Mat frame0, Mat frame1, Mat diff) {
        Core.absdiff(frame0, frame1, diff);
        for (int i = 0; i < 3; i++) {
            Imgproc.dilate(diff, diff, Imgproc.getStructuringElement(Imgproc.MORPH_RECT, new Size(3, 3)), new Point(0, 0), i + 1);
        }
        Imgproc.threshold(diff, diff, 30, 255, Imgproc.THRESH_BINARY);
    }

    private static void cannyEdgeDetection(Mat mat) {
        Imgproc.GaussianBlur(mat, mat, new Size(3, 3), 0);
        Imgproc.Canny(mat, mat, 120, 255, 3, true);
        Imgproc.dilate(mat, mat, Imgproc.getStructuringElement(Imgproc.MORPH_RECT, new Size(5, 5)), new Point(0, 0), 1);
    }

    private static List<Rect> getBoxesFromContours(
            List<MatOfPoint> contours, int originalWidth, int originalHeight, int resizedWidth, int resizedHeight) {
        List<Rect> boxes = new ArrayList<>();
        for (MatOfPoint contour : contours) {
            MatOfPoint2f contour2f = new MatOfPoint2f(contour.toArray());
            MatOfPoint2f approxCurve = new MatOfPoint2f();
            double approxDistance = Imgproc.arcLength(contour2f, true) * 0.02;
            Imgproc.approxPolyDP(contour2f, approxCurve, approxDistance, true);
            MatOfPoint points = new MatOfPoint(approxCurve.toArray());
            org.opencv.core.Rect rect = Imgproc.boundingRect(points);
            if (rect.width * rect.height
                    < 10000 * (resizedWidth * resizedHeight) / (originalWidth * originalHeight)) {
                continue;
            }
            Rect location = new Rect(
                    rect.x * originalWidth / resizedWidth,
                    rect.y * originalHeight / resizedHeight,
                    rect.x + rect.width * originalWidth / resizedWidth,
                    rect.y + rect.height * originalHeight / resizedHeight
            );
            if (location.left < location.right && location.top < location.bottom) {
                boxes.add(location);
            }
            contour2f.release();
            approxCurve.release();
            points.release();
        }
        return boxes;
    }
}

File: STRMCPP/src/main/cpp/src/RoIExtractor.cpp
#include "strm/RoIExtractor.hpp"

#include "opencv2/video/tracking.hpp"

#include "strm/Log.hpp"

namespace rm {

bool RoIExtractor::useOpticalFlowRoIs() const {
  return mConfig.OF_ROI;
}

void
RoIExtractor::process(const std::pair<std::pair<Frame*, Frame*>, std::vector<BoundingBox>>& item) {
  LOGD("RoIExtractor::process");
  Frame* prevFrame = item.first.first;
  Frame* currFrame = item.first.second;

  std::vector<RoI> rois;
  if (mConfig.OF_ROI) {
    std::vector<BoundingBox> prevResults;
    for (const BoundingBox& bbx : item.second) {
      if (bbx.confidence > mConfig.OPTICAL_FLOW_ROI_CONFIDENCE_THRESHOLD) {
        prevResults.push_back(bbx);
      }
    }
    std::vector<RoI> opticalFlowRoIs = getOpticalFlowRoIs(prevFrame, currFrame,
                                                          prevResults, mTargetSize);
    rois.insert(rois.end(), opticalFlowRoIs.begin(), opticalFlowRoIs.end());
    currFrame->opticalFlowRoIs = opticalFlowRoIs;
  }
  if (mConfig.PD_ROI) {
    std::vector<RoI> pixelDiffRoIs = getPixelDiffRoIs(prevFrame, currFrame, mTargetSize);
    rois.insert(rois.end(), pixelDiffRoIs.begin(), pixelDiffRoIs.end());
  }
  if (mConfig.MERGE_ROI) {
    mergeSingleFrameRoIs(currFrame, rois, mConfig.MERGE_THRESHOLD);
  }
  currFrame->rois = rois;
}

void RoIExtractor::mergeSingleFrameRoIs(
        const Frame* frame, std::vector<RoI> rois, float mergeThreshold) {
  while (true) {
    bool updated = false;
    for (auto it0 = rois.begin(); it0 != rois.end(); it0++) {
      for (auto it1 = it0 + 1; it1 != rois.end(); it1++) {
        const RoI& roi0 = *it0;
        const RoI& roi1 = *it1;
        float intersection = 0.0; // TODO: add STRMUtils.box_intersection(roi0.location, roi1.location);
        if (intersection / (float) roi0.getArea() < mergeThreshold
            && intersection / (float) roi1.getArea() < mergeThreshold) {
          continue;
        }
        int newLeft = std::min(roi0.location.left, roi1.location.left);
        int newTop = std::min(roi0.location.top, roi1.location.top);
        int newRight = std::max(roi0.location.right, roi1.location.right);
        int newBottom = std::max(roi0.location.bottom, roi1.location.bottom);
        if (newLeft >= newRight || newTop >= newBottom) {
          continue;
        }
        Rect newLocation(newLeft, newTop, newRight, newBottom);
        RoI::Type roiType = roi0.type == RoI::Type::OF || roi1.type == RoI::Type::OF
                            ? RoI::Type::OF
                            : RoI::Type::PD;
        std::string roiLabel = roi0.labelName.empty() || roi1.labelName.empty()
                               || roi0.labelName != roi1.labelName
                               ? "" : roi0.labelName;
        rois.erase(it0);
        rois.erase(it1);
        rois.emplace_back(frame, newLocation, roiType, roiLabel,
                          std::min(std::max(roi0.location.width(), roi0.location.height()),
                                   std::max(roi1.location.width(), roi1.location.height())));
        updated = true;
        break;
      }
      if (updated) {
        break;
      }
    }
    if (!updated) {
      break;
    }
  }
}

std::vector<RoI> RoIExtractor::getOpticalFlowRoIs(
        const Frame* prevFrame, const Frame* currFrame,
        const std::vector<BoundingBox>& boundingBoxes, const cv::Size& targetSize) {
  int width = currFrame->mat->cols;
  int height = currFrame->mat->rows;

  std::vector<Rect> boundingRects;
  boundingRects.reserve(boundingBoxes.size());
  for (const auto& bbx : boundingBoxes) {
    boundingRects.emplace_back(bbx.location);
  }

  std::vector<RoI> opticalFlowRoIs;
  if (!boundingBoxes.empty()) {
    const std::vector<std::pair<int, int>>& shifts = getBoundingBoxShifts(
            prevFrame->mat, currFrame->mat, boundingRects, targetSize);
    for (int boxIndex = 0; boxIndex < boundingBoxes.size(); boxIndex++) {
      const std::pair<int, int>& shift = shifts.at(boxIndex);
      const BoundingBox& box = boundingBoxes.at(boxIndex);
      const Rect& loc = box.location;
      int newLeft = std::max(0, loc.left + shift.first);
      int newTop = std::max(0, loc.top + shift.second);
      int newRight = std::min(width, loc.right + shift.first);
      int newBottom = std::min(height, loc.bottom + shift.second);
      if (newLeft < newRight && newTop < newBottom) {
        opticalFlowRoIs.emplace_back(
                currFrame, Rect(newLeft, newTop, newRight, newBottom),
                RoI::Type::OF, box.labelName);
      }
    }
  }
  return opticalFlowRoIs;
}

std::vector<std::pair<int, int>> RoIExtractor::getBoundingBoxShifts(
        const cv::Mat* prevImage, const cv::Mat* currImage,
        const std::vector<Rect>& boundingBoxes, const cv::Size& targetSize) {
  assert(prevImage != nullptr && currImage != nullptr);
  cv::Mat prevMat = prevImage->clone();
  cv::Mat currMat = currImage->clone();

  std::vector<cv::Point2f> p0;
  std::vector<cv::Point2f> p1;

  std::vector<uchar> status;
  std::vector<float> err;

  cv::cvtColor(prevMat, prevMat, cv::COLOR_BGR2GRAY);
  cv::cvtColor(currMat, currMat, cv::COLOR_BGR2GRAY);

  cv::resize(prevMat, prevMat, targetSize);
  cv::resize(currMat, currMat, targetSize);

  std::vector<cv::Point> centroids;
  for (const Rect& bbx : boundingBoxes) {
    int bbxCenterX = bbx.left + bbx.width() / 2;
    int bbxCenterY = bbx.top + bbx.height() / 2;
    cv::Point bbxCentroidPoints(bbxCenterX * targetSize.width / currImage->cols,
                                bbxCenterY * targetSize.height / currImage->rows);
    centroids.push_back(bbxCentroidPoints);
    // might not work... replaces p0.fromList(centroids); p0 is Point2f,
    p0.push_back(bbxCentroidPoints);
  }
  cv::TermCriteria criteria(cv::TermCriteria::COUNT + cv::TermCriteria::EPS, 10, 0.03);
  cv::calcOpticalFlowPyrLK(prevMat, currMat, p0, p1, status, err, cv::Size(15, 15), 2, criteria);

  uchar StatusArr[status.size()];
  std::copy(status.begin(), status.end(), StatusArr);

  cv::Point p1Arr[p1.size()];
  std::copy(p1.begin(), p1.end(), p1Arr);

  std::vector<std::pair<int, int>> shifts;
  for (int pointIdx = 0; pointIdx < centroids.size(); pointIdx++) {
    if (StatusArr[pointIdx] == 1) {
      shifts.emplace_back((int) ((p1Arr[pointIdx].x - centroids.at(pointIdx).x) * currImage->cols / targetSize.width),
                          (int) ((p1Arr[pointIdx].y - centroids.at(pointIdx).y) * currImage->rows / targetSize.height));
    } else {
      shifts.emplace_back(0, 0);
    }
  }
  return shifts;
}

std::vector<RoI> RoIExtractor::getPixelDiffRoIs(const Frame* prevFrame, const Frame* currFrame,
                                                const cv::Size& targetSize) {
  cv::Mat prevMat = prevFrame->mat->clone();
  cv::Mat currMat = currFrame->mat->clone();

  cv::cvtColor(prevMat, prevMat, cv::COLOR_BGR2GRAY);
  cv::cvtColor(currMat, prevMat, cv::COLOR_BGR2GRAY);

  cv::resize(prevMat, prevMat, targetSize);
  cv::resize(currMat, currMat, targetSize);

  cv::Mat mat;
  calculateDiffAndThreshold(prevMat, currMat, mat);
  cannyEdgeDetection(mat);

  std::vector<std::vector<cv::Point>> contours;
  cv::Mat hierarchy;

  cv::findContours(mat, contours, hierarchy, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

  // replaces get boxes from contours.
  std::vector<Rect> boxes;
  for (const std::vector<cv::Point>& contour : contours) {
    cv::Rect box = cv::boundingRect(contour);
    boxes.emplace_back(box.x, box.y, box.x + box.width, box.y + box.height);
  }

  std::vector<RoI> rois;
  rois.reserve(boxes.size());
  for (Rect& box : boxes) {
    rois.emplace_back(currFrame, box, RoI::PD, "");
  }
  return rois;
}

void RoIExtractor::calculateDiffAndThreshold(
        const cv::Mat& frame0, const cv::Mat& frame1, cv::Mat& diff) {
  cv::absdiff(frame0, frame1, diff);
  for (int i = 0; i < 3; i++) {
    cv::dilate(diff, diff, cv::getStructuringElement(
            cv::MORPH_RECT, cv::Size(3, 3)), cv::Point(0, 0), i + 1);
  }
  cv::threshold(diff, diff, 30, 255, cv::THRESH_BINARY);
}

void RoIExtractor::cannyEdgeDetection(cv::Mat& mat) {
cv::GaussianBlur(mat, mat, cv::Size(3, 3),
0);
cv::Canny(mat, mat,
120, 255, 3, true);
cv::dilate(mat, mat, cv::getStructuringElement(
        cv::MORPH_RECT, cv::Size(
5,5)), cv::Point(0,0), 1);
}

} // namespace rm
